### ISSUE TITLE
Reconnect to running session: keeping output

### ISSUE THREAD

**Issue by oersted** [jupyterlab/jupyterlab#2833]
There's been a lot of requests in the past to:
* Reconnect to a long-running notebook and keep receiving new output.
* Reconnect to a notebook running on the server side and recover all output when reconnecting.

I'd like to resurface this issue. To the best of my knowledge, these features have been planned at least since 2015 (earliest reference I found: [link](https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/jupyter/8hyGKoBY6O0/RyEDDyOZAQAJ)). It hasn't been implemented yet because it directly conflicts with the architecture of the stack and would require a big refactor.

I don't know if this is easier in the JupyterLab context or not. But now that this project is in very active development, it might be a good opportunity to revisit this and try to get it in now at the foundations.

I found to open issues on the topic in the old notebooks repo:
* jupyter/notebook#1150
* jupyter/notebook#641

**Comment by grahamanderson** [comment]
Is there some kind of alternative?
For myself, I have long running (data-science-y and scraping) Jupyter notebooks running (remotely) on a GCloud compute instance.
Kind of defeats the purpose if the script halts the moment I close my macbook. Or, am I mistaken? 
On my end, when I lose browser connection, I can't  reconnect to the Jupyter notebook without halting the script and losing output
Is there a plausible work-around?

**Comment by vidartf** [comment]
You could make a python client ("headless frontend") that auto-saves when new messages are received, and automatically shuts down when last pending execution is finished. If you spin this up in a separate process (from the notebook) before starting your execution, I think you should be able to do what you want. But this is not really a JupyterLab issue.

**Comment by ZmeiGorynych** [comment]
So am I correct in understanding that when you close the browser, the notebook keeps running, there is simply no way to connect to it again to view the output? If so, the logical thing to do is to have it dump summaries to file every so often, and use another, quick-to-restart notebook to display these.

**Comment by blink1073** [comment]
Hi @grahamanderson, I think the [mailing list](https://groups.google.com/forum/#!forum/jupyter) would be the best forum.

**Comment by Sarnath** [comment]
so, will this issue be addressed? Its so irritating when you have no idea what is happening on your long-running notebook. :(

**Comment by blink1073** [comment]
@Sarnath, this will be implemented when we implement real time collaboration, which is in [progress](https://github.com/jupyterlab/jupyterlab/issues/5382).

**Comment by SaschaHeyer** [comment]
@blink1073 any updates?

**Comment by blink1073** [comment]
The RTC work is ongoing and being tracked in https://github.com/jupyterlab/jupyterlab/issues/5382.  Adding the server side model is an extra step after that groundwork is laid.

**Comment by cossio** [comment]
> The RTC work is ongoing and being tracked in #5382. Adding the server side model is an extra step after that groundwork is laid.

Awesome! Looking so much forward to this issue being fixed.

**Comment by aseedb** [comment]
I am desperately waiting for this issue to be fixed. Please please fix it :-)

**Comment by hoangmt** [comment]
I have been reading all the related issues but there is no solution proposed. I really need this function.

**Comment by jasongrout** [comment]
We too would love if people would help contribute toward solving this!

**Comment by konradsitarz** [comment]
I'll need this functionality and I'm willing to help. Can anyone give me insight where to start? :v:

**Comment by jasongrout** [comment]
I think first step is to see if the current output buffering in the notebook server is working in JupyterLab. If there is *one* client connected to a kernel, and that client disconnects, then the server should be buffering output for when that one client reconnects. To test this, have a kernel that is generating output, then disconnect and reconnect to that kernel and see if output is buffered and sent to the client. There should be some messages in the server log about the buffering as well.

**Comment by eldad-a** [comment]
Thanks to all the Devs for their continuous contribution!
This scientific ecosystem is so helpful.

There is a partial work-around, which I re-post here for those who may find it helpful; it does not resolve the matter, yet provides some of the desired functionality.
Complemented with a log-file which is continuously updated, one can monitor the progress of the execution.

> There is a "half" of a work-around based on the [ipycache extension](https://github.com/rossant/ipycache).
Well, it may not be even a "half" but it is the best I found so far:
A cell can be "cached" such that running it again would load the output instead of executing the code again.

>For example:
After executing a cell containing

``` python
%load_ext ipycache
%mkdir -vp ipycache
import time
```

>A cell can be "decorated" using the `%%cache` magic:

``` python
%%cache ipycache/test_long 
i = 1
while i<30:
    print '{}, '.format(i),
    time.sleep(1)
    i+=1
```

>The browser can be closed 
>Running the last cell again will not execute it again but load the cached output (and display it).
>However, one has to wait for the execution to complete.
So this is will not help in monitoring the progress of a running code and I am not sure how it deals with errors.

_Originally posted in https://github.com/ipython/ipython/issues/4140#issuecomment-37864132_

**Comment by hamishcraze** [comment]
Hey guys, possibly a dumb suggestion, but why not just send all the output of a running cell to the first cell. If the issue is that cells lose which output maps to them, can't you just have a work around like:

"Insert generic message about why this is happening"
"Hey we see you reconnected to a running terminal, rerouting all output to first cell"

I mean I'm getting the messages of output under websocket traffic so the session is still receiving output. 
That seems like an easy work around. Or am I missing something nuanced about the problem?

**Comment by saulshanabrook** [comment]
Just a note that the current work on a real time data model is meant to address this issue as well: https://github.com/jupyterlab/rtc

**Comment by SolarisYan** [comment]
I'm suer the notebook server will buffer the output messages when client disconnet with the websocket, and server will replay the buffered messages to the websocket, but Jupyterlab don't update the cell output now. [Please refer to the notebook code](https://github.com/jupyter/notebook/blob/master/notebook/services/kernels/handlers.py#L256)

I think the key point is that clients every reconnect to the websocket, the session_id will change, so the buffered message can't map to the previous session.

wss://xxx/api/kernels/507fd6ea-ae93-441f-a902-3104cd46c3ae/channels?session_id=**9072fe45-23fb-4b30-9c28-8b5502264db0**&token=c63fb2b8268d4d949c070f3f6241f5f2
wss://xxx/api/kernels/507fd6ea-ae93-441f-a902-3104cd46c3ae/channels?session_id=**62cf6ab3-81db-40c3-8c6d-226c11759f02**&token=c63fb2b8268d4d949c070f3f6241f5f2

**Comment by echarles** [comment]
Very interesting issue. A few random thoughts:

- As @SolarisYan describes it well, I see the jupyterlab client initiating new sessions (new session_id) with my local server. I can understand a reconnect is needed if you network connection is flaky... But here, I am on my laptop, so the network is stable.
- I have not read anything on session drop/reconnect on https://github.com/jupyterlab/jupyterlab/blob/master/packages/services/README.md nor https://jupyter-client.readthedocs.io/en/stable/messaging.html but I may have read too fast.
- The token param will disappear thx to https://github.com/jupyterlab/jupyterlab/pull/8835 I guess.
- For RTC I guess @saulshanabrook you are thinking to the ability for a client to continue working offline in case of network issue, and then resubmit the changes that would be merged in a concurrent way with potential other user changes.

The above does not really help for the output keeping, but I wonder why I see so many session reconnection on my laptop. Is this random or is there a pattern for this?

**Comment by saulshanabrook** [comment]
> For RTC I guess @saulshanabrook you are thinking to the ability for a client to continue working offline in case of network issue, and then resubmit the changes that would be merged in a concurrent way with potential other user changes.

At least to be able to reconnect and get any changes made by the kernel... The offline work is a nice to have at this point but not dealbreaker for me!

**Comment by SolarisYan** [comment]
I'm sure i can have receive the replayed buffer messages from websocket after i reopen the notebook from left running kernel seesions, but the cell output don't change, i don't know why

**Comment by GF-Huang** [comment]
So what's the progress?

**Comment by ZhongxuanWang** [comment]
If you are interested in using magic cells commands, try 

`%%capture var`

Then, by default stdeer, stdout would be directed to the variable var

To reveal the content, simply

`var.show()`

Learn more about capture: https://ipython.readthedocs.io/en/stable/interactive/magics.html#cellmagic-capture

**Comment by jtlz2** [comment]
Sorry to ask, but what is the status of RTC now (with a view to enabling retention of output after a browser refresh)?

**Comment by mdbecker** [comment]
> @Sarnath, this will be implemented when we implement real time collaboration, which is in [progress](https://github.com/jupyterlab/jupyterlab/issues/5382).

It looks like #5382 is closed. Has this issue already been addressed?

**Comment by jasongrout** [comment]
The first version of RTC will ship with 3.1 soon, and paves the way to addressing this issue in the future.

**Comment by alikaanguven** [comment]
If you want to monitor output in realtime, you can write these two lines first:

`import sys`
`sys.stdout = open('my_log.log', 'w')`

And after your code, add this one to close:

`sys.stdout.close()`

Then you can simply monitor it from bash. You don't need to wait for the cell to be finished.

**Comment by cossio** [comment]
Now that RTC was released, are there any updates on this issue?

**Comment by williamstein** [comment]
Just a little "for what it's worth", I've implemented a few notebooks that *don't* have this problem, e.g., Sage Notebook, Sage worksheets in Cocalc, and CoCalc's Jupyter notebooks... and am happy to answer any questions.

**Comment by dmonad** [comment]
@williamstein I assume that you write the changes directly to the server and not to the client?

One possible solution is to write the output directly to the shared model so we don't need a client anymore to save the output to the server. Is there an easier way to implement this?

**Comment by williamstein** [comment]
> @williamstein I assume that you write the changes directly to the server and not to the client?

Yes.  In all cases there is code running in the same process as the server that takes the output messages from the kernel as input and constructs the ipynb file in memory (or something similar, e.g., keep images in sqlite as an optimization).

**Comment by jasongrout** [comment]
>  In all cases there is code running in the same process as the server that takes the output messages from the kernel as input and constructs the ipynb file in memory (or something similar, e.g., keep images in sqlite as an optimization).

IIRC, you started out with a model where output went to the server, and then was synced to the clients, but there were latency issues with that approach for the client executing the code. I recall that you then implemented a system where output would go directly to the executing client to have low latency, but then the logic for syncing between that client and the server was a bit trickier. Am I remembering correctly? How do things work now in Cocalc to keep latency low for the executing client, but still provide a consistent server-side record?

**Comment by sa-** [comment]
I would literally cancel my databricks subscription if this is fixed

**Comment by gogobd** [comment]
Are there any news on this issue? Workarounds?

**Comment by cossio** [comment]
Related: https://github.com/jupyter/notebook/issues/641

**Comment by blazespinnaker** [comment]
Workaround is to output to a file and echo it in jupyterlab.  Reconnecting would involve simply opening the file in python and printing it to the cell.  

Fairly straightforward, minor irritation it is not done automagically, but hardly a show stopper.  

Any thoughts on the best way to design this into a library?  maybe a magic?

**Comment by blazespinnaker** [comment]
This seems to work:

**Comment by blazespinnaker** [comment]
This seems to work.  There was a mention of ipycache magic (scroll up), but it required installing.

```
from contextlib import contextmanager
import sys
@contextmanager

def stdout_redirector():
    class MyStream:
        def write(self, msg):
            with open("logfile", "a") as f:
                f.write(msg)
            prefix = 'augmented:' if msg.strip() else ''
            old_stdout.write( prefix + msg)
        def flush(self):
            old_stdout.flush()
    old_stdout = sys.stdout
    sys.stdout = MyStream()
    try:
        yield
    finally:
        sys.stdout = old_stdout
    
with stdout_redirector():
    print("Test")

with open("logfile", "r") as f:
    print(f.read())
```

**Comment by aircloud** [comment]
I reviewed this problem and looked at the Jupyter code. The main problem now is that after refreshing the page or opening a new page, the UUID of the websocket changes, which makes it impossible to receive the previous output. I feel that the following ideas may solve this problem:

1. Modify the JupyterLab Workspace mechanism to a unique UUID
   * Now JupyterLab Workspace IDs are not unique, and different browsers may also conflict.
   * Also need to add some logic to clean up workspace persistent files
2. Modify the UUID logic of JupyterLab Notebook connecting to the kernel, change it to `workspaceId-notebookId`, then it is unique.
3. Modify the msgID of the message to `workspaceId-notebookId-cellID`

Then we can add recovery logic based on the above points. After refreshing the page, write the output to the corresponding cell according to the UUID of the message

But I haven't fully verified it, and I'm not quite sure if this is in line with JupyterLab's design philosophy

**Comment by davidbrochart** [comment]
This issue is already fixed with RTC. You can try it by installing a pre-release of JupyterLab v4, e.g.:
```console
pip install jupyterlab==4.0.0a28
```
And enabling collaborative mode:
```console
jupyter lab --collaborative
```

**Comment by dlukes** [comment]
@davidbrochart Could you please specify what workflow you tried that worked w.r.t. reconnecting to the running session and keeping output? I tried running JupyterLab from the command line (using the version and invocation you specified), running the following cell...

```python
import time
while True:
    print("foo")
    time.sleep(1)
```

... then closing the JupyterLab browser tab and reopening it via the URL shown in the terminal. What I get is an idle session indicator (even though the session is actually busy) and no output being added below the cell.

**Comment by davidbrochart** [comment]
You're right.
What I tried was opening a window with the URL, running your code, then a private window with the URL. Both windows are in sync. Then closing the second window and reopening a private window with the URL. They are still in sync and the previous output was also shown.
But closing the first window (the one where the code was run) stops the output display in the second window. I remember now that it's the expected behavior, because at least one front-end has to render outputs. Sorry for saying that this was fixed in RTC.
Jupyverse has a [new execution mode](https://github.com/jupyter-server/jupyverse/pull/191) that would truly fix this issue, but as far as I know there is no front-end that takes advantage of it.

**Comment by davidbrochart** [comment]
There is already #12867.

**Comment by dlukes** [comment]
> Sorry for saying that this was fixed in RTC.

No worries, thanks for clarifying!

**Comment by pranavkrishna** [comment]
Is this issue fixed? I could help if not.

**Comment by krassowski** [comment]
Linking the current discussion: https://github.com/jupyter-server/jupyter_server/issues/1274

**Comment by davidbrochart** [comment]
I made some progress towards restoring notebook state, using jupyverse/jpterm:

https://github.com/jupyter-server/jupyter_server/assets/4711805/a446b3e6-c65d-40b4-bdf1-4951a4f793c3

**Comment by sa-** [comment]
@davidbrochart That's awesome! Is this done by storing a Ydoc in the kernel?

I'm imagining a setup with kernel gateway or enterprise gateway. It would be very nice to have jupyter running locally, and then ideally nothing would break if there is a power or internet outage during my 2 hour long cell execution

**Comment by davidbrochart** [comment]
There is a YDoc representing the notebook but not in the kernel. This YDoc lives in the (jupyverse) server and in the (jpterm) client.
I'm currently working on doing the same for widgets, and in this case yes, there will be a YDoc representing the widget in the kernel, in the server and in the client. The widget will synchronize between the kernel and the server using the Comm protocol, and between the server and the client using a WebSocket. It will allow widget state restore as well.

**Comment by Wh1isper** [comment]
It seems to go further with my comment [before](https://github.com/jupyter-server/jupyter_server/issues/1274#issuecomment-1657944667) that using `Y-CRDT`(`YDOC`) to establish consistency, we can develop a wide variety of applications. 👍

**Comment by sa-** [comment]
@davidbrochart that is great to hear. I would like to contribute but I'm new to the jupyter codebase - is there anything I can help with?

**Comment by jliebers** [comment]
@davidbrochart Your video looks amazing. I just want to applaud 👏 any effort towards resolving this issue, since it is the biggest disappointments that jupyterlab currently has (at least to me :) ). Any solution towards resolving this issue is highly welcome - thank you for your work in this direction!

**Comment by davidbrochart** [comment]
And here is a demo showing full notebook state recovery, including widgets:

https://github.com/jupyterlab/jupyterlab/assets/4711805/5625ae3b-c8e3-4b6a-8328-7c32cbf396eb

**Comment by TeaCult** [comment]
Wouldn't that be easier to keep a running session as an html document on the server side, and connect to it just to change this html. So that it would never disconnects only the renderer and editor of this html could disconnect. Basically I am trying to say that run an jupyter html in the server. And connect to it like google sheets(multiple users or single) and edit this html and trigger run and render output from this. Wouldn't it preserve states easily ?

**Comment by davidbrochart** [comment]
It doesn't work like this, there is no such thing as an HTML document on the server side.

**Comment by jianghuife** [comment]
> And here is a demo showing full notebook state recovery, including widgets:
> 
>  Peek.2023-11-10.11-29.mp4

How do I achieve this effect？

**Comment by davidbrochart** [comment]
If you are asking about how to do it in [jpterm](https://github.com/davidbrochart/jpterm), please open an issue there.
Otherwise, there is a [PR](https://github.com/jupyterlab/jupyterlab/pull/15448) in JupyterLab.

**Comment by krassowski** [comment]
Well, it is not yet implemented in jupyter-server, and not released but there definitely is some progress here ;)

**Comment by astrojuanlu** [comment]
xref https://github.com/jupyter-server/jupyter_server/issues/1274

**Comment by davidbrochart** [comment]
No, it means that this feature will be available if you run JupyterLab with [jupyverse](https://github.com/jupyter-server/jupyverse) instead of [jupyter-server](https://github.com/jupyter-server/jupyter_server), until it is implemented in `jupyter-server`.

**Comment by pengzhenghao** [comment]
@davidbrochart Could you please share how we can achieve this via jupyverse? I've installed jupyverse and still the output can't be seen after closed & reopen a browser tab.

**Comment by davidbrochart** [comment]
```bash
pip install "jupyter-collaboration" "jupyverse[jupyterlab,auth]"
jupyverse --set kernels.require_yjs=true --set jupyterlab.server_side_execution=true
```
